File: 2940-find-building-where-alice-and-bob-can-meet/2940-find-building-where-alice-and-bob-can-meet.java
import heapq

class Solution:
    def leftmostBuildingQueries(self, heights: list[int], queries: list[list[int]]) -> list[int]:
        n = len(heights)
        m = len(queries)
        v = [[] for _ in range(n)]
        ans = [-1] * m

        for i in range(m):
            x, y = queries[i]
            if x > y:
                x, y = y, x
            if x == y or heights[x] < heights[y]:
                ans[i] = y
            else:
                v[y].append((max(heights[x], heights[y]), i))

        pr = []
        for i in range(n):
            for it in v[i]:
                heapq.heappush(pr, it)
            while pr and pr[0][0] < heights[i]:
                ans[heapq.heappop(pr)[1]] = i

        return ans

File: 2940-find-building-where-alice-and-bob-can-meet/2940-find-building-where-alice-and-bob-can-meet.cpp
class Solution {
public:
    vector<int> leftmostBuildingQueries(vector<int>& heights, vector<vector<int>>& queries) {
        int n = heights.size(), m = queries.size();
        vector<vector<pair<int,int>>>v(n);
        vector<int>ans(m,-1);
        for(int i = 0 ; i < m ;i++){
            int x = queries[i][0],y = queries[i][1];
            if(x > y) swap(x,y);
            if(x == y || heights[x] < heights[y])ans[i] = y;
            else v[y].push_back({max({heights[x],heights[y]}),i});
        }

        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<>>pr;
        for(int i = 0 ; i < n;i++){
            for(auto it : v[i])pr.push(it);
            while(pr.size() && pr.top().first < heights[i]){
                ans[pr.top().second] = i;
                pr.pop();
            }
        }
        return ans;
    }
};